Better Control over Concurrency

ArquintL opened the pull request:
This PR adds the possibility to pass an execution context consisting of a thread pool and actor system to ViperCoreServer

ArquintL commented on src/test/scala/CoreServerTests.scala:
```diff
@@ -64,6 +64,10 @@ class CoreServerTest extends WordSpec with Matchers with ScalatestRouteTest {
 
+  // create a new execution context for each ViperCoreServer instance
+  // (otherwise there is a not-unique exception occurring for the Terminator actor when concurrently verifying multiple programs
+  private implicit def getExecutionContext: VerificationExecutionContext = new DefaultVerificationExecutionContext()
```
@aterga do you might have an idea why reusing the same execution context / actor system does not work (i.e. when changing this function definition to a `val`?

aterga replied:
The comment above was added by Valentin, I think. Perhaps it's some simple issue, like two instances of the Terminator actor would occur in the same system with the same name? If that's the case, one could run multiple servers inside one actor system by making sure that we add some kind of unique server identifier as part of the name of all actors that we create. (Two actors wit the same name cannot exist in one actor system.)

aterga replied:
I would just try running any see what would fail.

aterga commented on src/main/scala/viper/server/core/DefaultVerificationExecutionContext.scala:
```diff
@@ -0,0 +1,41 @@
+object DefaultVerificationExecutionContext {
+  // at least 2 threads are needed if the actor system is also using the same thread pool (as opposed to creating a
+  // actor system with the default executor like `ActorSystem(actorSystemName)`)
+  val minNumberOfThreads: Int = 2
```
Could it be that this constant is different if the server creates instances of Carbon and Silicon? Silicon uses some concurrency so it may affect the thread pool, whereas Carbon, as far as I remember, does not.

ArquintL replied:
The execution context is not passed to the backend (not to increase the range of my changes even further). Hence, Silicon uses it's own thread pool and I think this constant is thus independent of the chosen backend

aterga commented on src/main/scala/viper/server/core/DefaultVerificationExecutionContext.scala:
```diff
@@ -0,0 +1,41 @@
+  val minNumberOfThreads: Int = 2
+}
+
+class DefaultVerificationExecutionContext(actorSystemName: String = "Actor_System", threadNamePrefix: String = "thread") extends VerificationExecutionContext {
```
Could you please add a Scaladoc to this file to explain, in 1-2 paragraphs, what is the motivation for this class, and what are the intended use cases?

aterga commented on src/main/scala/viper/server/vsi/VerificationServer.scala:
```diff
@@ -170,16 +169,15 @@ trait VerificationServer extends Unpacker {
-    private def startJob(task: Thread, queue: SourceQueueWithComplete[Envelope], publisher: Publisher[Envelope]): JobHandle = {
-      _verificationTask = task
-      _verificationTask.start()
+    private def startJob(task: VerificationTask, queue: SourceQueueWithComplete[Envelope], publisher: Publisher[Envelope])(executor: VerificationExecutionContext): JobHandle = {
```
One could probably make ```executor``` an implicit parameter.

ArquintL replied:
One could but this in my understanding this would only affect the functions body in which I do not need an implicit executor (the call to `submit` requires an explicit receiver since Scala does not seem to accept an implicit receiver)

aterga commented on src/main/scala/viper/server/vsi/VerificationServer.scala:
```diff
@@ -323,14 +320,20 @@ abstract class VerificationTask()(implicit val executionContext: ExecutionContex
-    })
-    while(current_offer == null || !current_offer.isCompleted){
-      Thread.sleep(10)
+    Await.ready(answer, Duration.Inf)
```
This part would need to be adapted after PR22 is merged. I suspect that rather than waiting for the future to end we should just return it to the client. Working with futures is quite convenient because they can always be flattened, but flattening (and, moreover, await) should happen lazily.

ArquintL replied:
I agree that futures should be passed on the clients instead of awaiting them in ViperServer. However, I do not agree with you in this particular use since this function is concerned with enqueuing messages to the message queue. As we've discussed in our meetings, the verification job should be slowed down (i.e. blocked) when the queue is full. Or did I miss something?

aterga replied:
Okay, but then, I think, we need to Await only once, in the end of ```enqueueMessages```. 
How about something like this?: 

```scala
protected def enqueueMessages(msg: A): Unit = {
  implicit val askTimeout: Timeout = Timeout(5000 milliseconds)
  val answer = q_actor ? TaskProtocol.BackendReport(pack(msg))
  val current_offer = answer.map {
    case Success(current_offer) => //...
    // ... 
  }
  Await.result(current_offer, Duration.Inf)
}
```

ArquintL replied:
I changed it now to (newly located in `MessageStreamingTask`):
```
// answer is a future that will resolve with the actor's response to the BackendReport request
val answer = (q_actor ? TaskProtocol.BackendReport(msg)).mapTo[Future[QueueOfferResult]]
// currentOffer is the future that the actor will send in its response (assuming that no timeout occurred requesting it from the actor)
// currentOffer will resolve when the message is dequeued from the queue
val currentOffer = answer.flatten
try {
  // note that an exception is thrown if the currentOffer future fails, e.g. because the askTimeout occurred
  Await.result(currentOffer, Duration.Inf)
} catch {
  case ex: Exception =>
    // print exception such that one sees that something went wrong:
    println(s"exception in enqueueMessage occurred: $ex")
    // rethrow exception:
    throw ex
}
```

aterga replied:
Generally looks reasonable now. As a recommendation for future, ```Duration.Inf``` should be avoided in favour of fault tolerance (maybe worth adding a TODO).

aterga reviewed (changes requested):
Thanks for the PR! I've added some comments.

aterga reviewed (changes requested):
Responded to some questions.

ArquintL commented:
Copying the following test case to `AsyncCoreServerSpec` in master branch results in the same issues as on this branch:
```
private val ver_error_file = "src/test/resources/viper/verification_error.vpr"
private val empty_viper_file = "src/test/resources/viper/empty.vpr"
private val correct_viper_file = "src/test/resources/viper/sum_method.vpr"
private val files = List(empty_viper_file, correct_viper_file, ver_error_file)
private val asts = files.map((new AstGenerator(SilentLogger().get).generateViperAst(_).get))
private def getAstByFileName(file: String): Program =
  (files zip asts collect {
    case (f, ast) if f==file => ast
  }).last
it should s"be able to verify multiple programs with caching disabled and retrieve results" in {
  val jobIds = files.map(file => (file, core.verify(file, silicon_without_caching, getAstByFileName(file))))
  val filesAndMessages = jobIds map { case (f, id) => (f, ViperCoreServerUtils.getMessagesFuture(core, id)) }
  val resultFutures = filesAndMessages map { case (f, fut) => fut.map(msgs => {
    println(s"future completed for $f")
    msgs.last match {
      case _: OverallSuccessMessage => assert(f != ver_error_file)
      case _: OverallFailureMessage => assert(f == ver_error_file)
      case msg => fail(s"unexpected message: $msg")
    }
  })}
  Future.sequence(resultFutures).map(results => assert(results.length == 3))
}
```

ArquintL commented:
@aterga I've merged your latest changes into this branch.
Except the obvious addition of the VerificationExecutionContext that provides an actor system and execution context, I've changed `AstWorker`, `VerificationWorker`, `MessageStreamingTask`, and `MessageReportingTask` to no longer inherit from Thread. I find it more natural that these concepts inherit from Callable[T] and return a result of type T. They provide a FutureTask that can be submitted to the thread pool and still have the artifact future. Also different from your changes is the `VerificationWorker` that no longer uses type parameter `AST` but `Unit` as it does not provide an artifact. Imho these changes make sense to me but if you disagree I'm curious how you'll convince me ;) 

In addition, I've migrated all CoreServerTests to their async counterpart. However, you'll see that simultaneously verifying multiple files does not seem to work. I've added corresponding TODOs and restricted the tests to a single file (as this is working). I've tried running it on the master branch and the same issue seems to occur there as well (see my comment above). I think the cause is in Silicon and/or Silver and might even be related to @mschwerhoff's performance issues. I don't know yet whether I'll have time before the end of January to further look into this but would you agree that the next step is to directly reproduce it with Silicon?

Could you please have a look at this PR again?

aterga commented:
The summary of your changes sounds good to me. 

> Also different from your changes is the VerificationWorker that no longer uses type parameter AST but Unit as it does not provide an artifact. 

That's fine, I did not attach any special meaning to the VerificationWorker's AST artifact. FYI: Akka also provides ```Done``` s.t. one can return e.g. ```Future[Done]``` instead of ```Future[Unit]``` which, in some cases, makes the intention of the code clearer. 

> I've tried running it on the master branch and the same issue seems to occur there as well (see my comment above). I think the cause is in Silicon and/or Silver and might even be related to @mschwerhoff's performance issues. I don't know yet whether I'll have time before the end of January to further look into this but would you agree that the next step is to directly reproduce it with Silicon?

Sounds reasonable!

aterga reviewed (approved):
This PR looks good to me — thanks for working it out! Feel free to merge.

ArquintL commented:
An update on the failing test cases:
- The one test case on wrong positional information would be fixed by @marcoeilers's [workaround](https://github.com/viperproject/silver/pull/494) (and by calling `resetCache()` in `ViperCache.initialize(...)`
- Simultaneously parsing multiple Viper files seems to be still broken even with @fpoli's [fix](https://github.com/viperproject/silver/pull/493). I tried to use commits from October to test whether it was working back then but to many changes are necessary in ViperServer to get it compiling with these older versions of Silver, Silicon, and Carbon. I think time will be better spent on debugging AST rewriting and figuring out why it does not terminate after the paper deadline.

fpoli commented:
> I think time will be better spent on debugging AST rewriting and figuring out why it does not terminate after the paper deadline.

If you modify the AST while iterating over its nodes, then [this commit](https://github.com/viperproject/silver/commit/1df0f34d2be400be42a32a599dad6ecef12f1315#diff-4d40416248f1a65d2eeef45692b1fb1eeee65a2746204806ac4afd264da5e813R64-R80) changes the overall behavior and might be the cause of the failing tests. PR https://github.com/viperproject/silver/pull/493 just improves the performance and does not change this behavior. (I suggest to add a test case with simultaneous parsing in Silver, if possible.)

fpoli commented:
> > I think time will be better spent on debugging AST rewriting and figuring out why it does not terminate after the paper deadline.
> 
> If you modify the AST while iterating over its nodes, then [this commit](https://github.com/viperproject/silver/commit/1df0f34d2be400be42a32a599dad6ecef12f1315#diff-4d40416248f1a65d2eeef45692b1fb1eeee65a2746204806ac4afd264da5e813R64-R80) changes the overall behavior and might be the cause of the failing tests. PR [viperproject/silver#493](https://github.com/viperproject/silver/pull/493) just improves the performance and does not change this behavior. (I suggest to add a test case with simultaneous parsing in Silver, if possible.)

I updated the PR https://github.com/viperproject/silver/pull/493 to restore the old behavior. So, it might fix bugs.
